### PR TITLE
Make main task queue size() a constant-time operation

### PR DESCRIPTION
**Pull request by vemacs**
I don't think contention is an issue as I am using LongAdder. This solves many scalability issues with main thread tasks in general (especially network ones at high player counts).

### CODE REVIEW

**Review by AlfieC — APPROVED** [review]

**Review by AlfieC — APPROVED** [review]

File: Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ import javax.imageio.ImageIO;
+ import org.apache.commons.lang3.Validate;
+@@ -102,7 +103,33 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
+     private final GameProfileRepository X;
+     private final UserCache Y;
+     private long Z;
+-    protected final Queue<FutureTask<?>> j = new java.util.concurrent.ConcurrentLinkedQueue<FutureTask<?>>(); // Spigot, PAIL: Rename
++    // Paper - Make size() constant-time
++    protected final Queue<FutureTask<?>> j = new java.util.concurrent.ConcurrentLinkedQueue<FutureTask<?>>() {
++        private LongAdder cachedSize = new LongAdder();
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch]: Why is cachedSize not final?

**Comment by Spottedleaf** [comment]
Do we really need to use CLQ.size()?

File: Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch (hunk 1 above)
Comment (vemacs) [reply Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch]: Fixing, rebuild is unreliable for me

**Comment by vemacs** [comment]
I'm not sure to be honest, but it's best to just retain similar behavior. We can ship this, compare the results, and then evaluate if the check can be further optimized. Don't let perfect be the enemy of good :^)

File: Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch (hunk 1 above)
Comment (vemacs) [reply Spigot-Server-Patches/0180-Make-main-task-queue-size-a-constant-time-operation.patch]: Pushed fixed patch

**Comment by Spottedleaf** [comment]
Alright.

Any data to report significant gain? I seem to recall you stating that you had such data.

**Comment by vemacs** [comment]
@AlfieC has some 1.7 numbers, I'm trying to set up pycraft to get 1.11 numbers. Somewhere in the range of a 5ms decrease in per-tick time at 200 players.

**Comment by AlfieC** [comment]
The reason that size() is used is because the runnables aren't free to run and take time - during which tasks can be forever added which mean a tick could end up lasting forever.

**Comment by AlfieC** [comment]
However, it isn't a huge issue and could be eliminated. But it's best to retain old behaviour and such to make maintaining diffs easier.

**Comment by AlfieC** [comment]
With this and a few rudimentary optimisations, I was able to reduce tick time to sub 0.1ms with over 100 active, non-bot players.

**Comment by aikar** [comment]
This looks fine to me, but can you move it to a named class instead so the diff is simpler.

**Comment by aikar** [comment]
```The reason that size() is used is because the runnables aren't free to run and take time - during which tasks can be forever added which mean a tick could end up lasting forever.```

The wording you gave even confused me, and but in simpler terms, its to make sure the server doesnt process new tasks added after it started processing, as the server is always receiving packets, and without a limit, it could keep running until it finally hits a quiet period.

**Review by AlfieC — APPROVED** [review]

**Comment by AlfieC** [comment]
Better title would just be "Make ConcurrentLinkedQueue size() O(1)"
